### ISSUE TITLE
Functionality to Convert LangChain Tools to Mirascope Tools

### ISSUE THREAD

**Issue by minhquan23102000** [Mirascope/mirascope#639]
### Question

**Description:**

Currently, developers who want to use tools  within Mirascope need to reimplement them. This can be time-consuming and inefficient. It would be beneficial to have a feature or utility that automatically converts LangChain tools into Mirascope-compatible tools.

Langchain support wide range of tools and toolkit: https://python.langchain.com/docs/integrations/tools/

**Motivation:**

This feature would significantly simplify the development process for users who want to leverage the extensive collection of tools available in LangChain within the Mirascope framework. It would reduce development time and effort, allowing developers to focus on building their applications rather than reimplementing existing tools.

**Examples:**

* A function or script that takes a LangChain tool object as input and outputs a Mirascope-compatible tool object.
* A documentation page outlining the process of converting LangChain tools to Mirascope tools manually, if automatic conversion is not feasible.

**Benefits:**

* Reduced development time and effort.
* Increased accessibility of LangChain tools within Mirascope.
* Improved developer experience.

**Alternatives:**

Currently, the only alternative is to manually reimplement tools in Mirascope. This is a time-consuming and potentially error-prone process.

**Comment by willbakst** [comment]
There are two things here that I think are worth digging into:

__1. Pre-made Tools__

From your question/comment, it seems worthwhile for Mirascope to finally implement the `mirascope.tools` module with a bunch of pre-made ready-to-go tools for users so they can get up and running quickly without having to write their own tools.

Is there a list of specific tools you would like to see added to this module first? That way we can prioritize those tools in the upcoming work. You can also create individual GitHub Issue Feature Requests for any particular tools you want to see in the library and we (or other community members) can work to include them :)

__2. How to use existing LangChain tools with Mirascope__

To be honest, I looked into implementing at least something makeshift for conversion, but their tools are all structured and implemented differently (and poorly imo), so the conversion script would be quite time consuming (and break if they make any changes to any tools).

I would much rather focus on (1) and implement the `mirascope.tools` module / suite of pre-made tools for our community, as I think this will ultimately be much better in the long run.

In the meantime, here's the best way to manually use a LangChain tool with Mirascope:

1. Create a Mirascope `BaseTool` subclass
2. Instantiate the LangChain tool as a class variable of your Mirascope tool
3. Set `__name__` equal to the LangChain tool's `.__class__.__name__` value
4. Override the `_name` and `_description` class methods using the LangChain tool's name/description field values
5. Add the fields necessary for running the LangChain tool
6. Implement the `call` method simply by calling the LangChain tool with the necessary fields

Here's an example of using their `DuckDuckGoSearchRun` tool:

```python
from typing import ClassVar

from langchain_community.tools import DuckDuckGoSearchRun
from mirascope.core import BaseTool

class DuckDuckGoSearch(BaseTool):
    __name__ = langchain_tool.__class__.__name__

    langchain_tool: ClassVar[DuckDuckGoSearchRun] = DuckDuckGoSearchRun()

    query: str

    def call(self) -> str:
        return self.langchain_tool._run(self.query)

    @classmethod
    def _name(cls) -> str:
        return cls.langchain_tool.name

    @classmethod
    def _description(cls) -> str:
        return cls.langchain_tool.description
```

For reference, I started implementing a utility for doing the above automatically but quickly realized that there was very little consistency between tools, so I didn't bother continuing further with that effort.

Hope this helps and looking forward to hearing what tools you want to see implemented in `mirascope.tools` first!

**Comment by willbakst** [comment]
Awesome, we've started work on `mirascope.tools` and plan to include some web tools (e.g. search, parse content, etc.) as well as some toolkits (e.g. file manager).

Once we release this initial module, I'll parse each of your additional suggestions into individual GitHub issues so we can discuss and implement each one individually :)

**Comment by willbakst** [comment]
Personally, I would rather have full control over the tools and implement as part of the `mirascope.tools` module (which we're already making good progress on and should have the initial release out this Friday 😃)

If you're interested in contributing some form of conversion functionality, let us know! We'd love to help with your contribution.

**Comment by willbakst** [comment]
These are great! We just released the initial `mirascope.tools` module in `v1.7`. Some of these tools you've written already exist in some form, but if you see a means of improving our implementations (or differences worth including as separate tools), by all means!

Please take a look at [what we've implemented](https://github.com/Mirascope/mirascope/tree/main/mirascope/tools) so you can get a sense of how we envision the future of the `mirascope.tools` module. For example, we've added functionality for each tool to have a usage prompt that can be included in the messages array (e.g. the system message) to instruct the LLM how to use it.

I think a great starting contribution would be a `WikipediaToolKit` that contains the tools you've shared.

We're currently working on the pre-made toolkit implementations, so it's likely best to wait until we get that [initial implementation](https://github.com/Mirascope/mirascope/pull/651) merged in so things are more settled for your contribution.

You can find our contribution guidelines [here](https://www.mirascope.com/docs/CONTRIBUTING/).

**Comment by willbakst** [comment]
`mirascope.tools` is released in `v1.8`!

I'm going to close this issue. Any additional work (e.g. new tools etc.) should be opened as separate issues so we can better track the individual requesets/work.

Thank you for suggesting we do this and helping to push mirascope to be the best version of itself it can be!
